packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2550,14 +2545,13 @@ export class ContainerRuntime
 			// replay the ops
 			this.pendingStateManager.replayPendingStates();
 		} finally {
-			// Save the new start and restore the old state, re-enable event emit
-			newState = this.dirtyContainer;
-			this.dirtyContainer = oldState;
+			// Restore the old state, re-enable event emit
+			this.lastEmittedDirty = oldState;
 			this.emitDirtyDocumentEvent = true;
 		}
 
-		// Officially transition from the old state to the new state.
-		this.updateDocumentDirtyState(newState);
+		// This will emit an event if the state changed relative to before replay
+		this.updateDocumentDirtyState();
```

Comment:
Potential behavior change here (in corner cases): This will recompute dirty state rather than using the "last emitted" state from during the replay.

So this will be another case where if the only pending message is non-dirtyable, the state will say "dirty" when it could/should say "saved".  Thinking through non-dirtyable ops, I don't think this is an issue. Possibly with Agent Scheduler, I'm not familiar with that as much.

But worst case would be a spurious "unsaved changes" dialog after successful reconnect, until the op roundtrips.  Summarizer also checks isDirty state, but Summarizer closes on reconnect so this path isn't a concern.

Reply:
Fixed - Updated the "dirty" computation to account for these all the time.